ztools/ztls/handshake_client.go

```diff
@@ -32,19 +32,246 @@ type clientHandshakeState struct {
+	HandshakeVersion uint16
+
+	//if len == 32, it will specify the client random
+	//Otherwise, the field will be random
+	ClientRandom []byte
+
+	//if RandomSessionID > 0, will overwrite SessionID w/ that many
+	//random bytes when a session resumption occurs
+	RandomSessionID int
+	SessionID       []byte
+
+	//These fields will appear exactly in order in the ClientHello
+	CipherSuites       []uint16
+	CompressionMethods []uint8
+	Extensions         []ClientExtension
+
+	//Optional, both must be non-nil, or neither.
+	//Custom Session cache implementations allowed
+	SessionCache ClientSessionCache
+	CacheKey     CacheKeyGenerator
+}
+
+func (c *ClientHelloConfiguration) ValidateExtensions() error {
+	for _, ext := range c.Extensions {
+		switch ext.(type) {
+		case PointFormatExtension:
+			for _, format := range ext.(PointFormatExtension).Formats {
+				if format != pointFormatUncompressed {
+					return errors.New(fmt.Sprintf("Unsupported EC Point Format %d", format))
+				}
+			}
+		case SignatureAlgorithmExtension:
+			for _, algs := range ext.(SignatureAlgorithmExtension).getStruct() {
+				found := false
+				for _, supported := range supportedSKXSignatureAlgorithms {
+					if algs.hash == supported.hash && algs.signature == supported.signature {
+						found = true
+						break
+					}
+				}
+				if !found {
+					return errors.New(fmt.Sprintf("Unsupported Hash and Signature Algorithm (%d, %d)", algs.hash, algs.signature))
+				}
+			}
+		}
+	}
+	return nil
+}
+
+func (c *ClientHelloConfiguration) ModifyConfig(config *Config) *Config {
+	config.NextProtos = []string{}
+	config.CipherSuites = c.CipherSuites
+	config.MaxVersion = c.HandshakeVersion
+	config.ClientRandom = c.ClientRandom
+	config.CurvePreferences = []CurveID{}
+	config.HeartbeatEnabled = false
+	config.ExtendedRandom = false
+	config.ForceSessionTicketExt = false
+	config.ExtendedMasterSecret = false
+	config.SignedCertificateTimestampExt = false
+	for i, ext := range c.Extensions {
+		switch ext.(type) {
+		case SniExtension:
+			if ext.(SniExtension).Autopopulate {
+				c.Extensions[i] = SniExtension{[]string{config.ServerName}, true}
+			}
+			if config.ServerName == "" && len(ext.(SniExtension).Domains) > 0 {
+				config.ServerName = ext.(SniExtension).Domains[0]
+			}
+		case ALPNExtension:
+			config.NextProtos = ext.(ALPNExtension).Protocols
+		case ExtendedMasterSecretExtension:
+			config.ExtendedMasterSecret = true
+		case SignatureAlgorithmExtension:
+			supportedSKXSignatureAlgorithms = ext.(SignatureAlgorithmExtension).getStruct()
+			defaultSKXSignatureAlgorithms = ext.(SignatureAlgorithmExtension).getStruct()
+		case SCTExtension:
+			config.SignedCertificateTimestampExt = true
+		case SupportedCurvesExtension:
+			config.CurvePreferences = ext.(SupportedCurvesExtension).Curves
+		case SessionTicketExtension:
+			config.ForceSessionTicketExt = true
+		}
+	}
+	return config
+}
+
+func (c *ClientHelloConfiguration) marshal(config *Config) ([]byte, error) {
```

Comment:
This is making the on-the-wire version of the `ClientHelloConfiguration`. I have this distinct from the `clientHelloMsg` serializer to make clearer that they are very different cases. It needs the `Config` to specify the random number generator, and to see if the config is specifying `ForceSuites`.